[philox_rand] Dynamic shape support

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99289
* __->__ #99290


Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this


~~~


[2023-04-24 21:41:37,446] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
 ===== Forward graph 1 =====
 <eval_with_key>.7 class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: i64[], arg1_1: i64[], arg2_1: Sym(s0), arg3_1: Sym(s1), arg4_1: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(arg1_1, 0)
        philox_rand = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        getitem: f32[s0, s1] = philox_rand[0]
        getitem_1: i64[] = philox_rand[1];  philox_rand = None
        add_1: i64[] = torch.ops.aten.add.Tensor(getitem_1, 0);  getitem_1 = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem, arg4_1);  getitem = arg4_1 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        add_2: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_1)
        philox_rand_1 = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add_2, None, device(type='cuda', index=0), torch.float32);  arg2_1 = arg3_1 = arg0_1 = add_2 = None
        getitem_2: f32[s0, s1] = philox_rand_1[0]
        getitem_3: i64[] = philox_rand_1[1];  philox_rand_1 = None
        add_3: i64[] = torch.ops.aten.add.Tensor(add_1, getitem_3);  add_1 = getitem_3 = None
        mul_1: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem_2, mul);  getitem_2 = mul = None

        # No stacktrace found for following nodes
        add_4: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_3);  arg1_1 = add_3 = None
        return (mul_1, add_4)

 ~~~

Each rand op is accompanied by its offset calculation op.

anijain2305 committed:
[philox_rand] Dynamic shape support

[ghstack-poisoned]

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"




[ghstack-poisoned]

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"




[ghstack-poisoned]

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this

~~~
 ===== Forward graph 1 =====
 <eval_with_key>.10 class GraphModule(torch.nn.Module):
    def forward(self, fwd_seed_1: i64[], fwd_base_offset_1: i64[], primals_1: Sym(s0), primals_2: Sym(s1), primals_3: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, 0)
        philox_rand: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand, primals_3);  primals_3 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        mul_1: Sym(s0*s1) = primals_1 * primals_2
        sub: Sym(s0*s1 - 1) = mul_1 - 1
        add_1: Sym(s0*s1 + 256) = mul_1 + 256;  mul_1 = None
        sub_1: Sym(s0*s1 + 255) = add_1 - 1;  add_1 = None
        floordiv: Sym(((s0*s1 + 255)//256)) = sub_1 // 256;  sub_1 = None
        mul_2: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv;  floordiv = None
        mul_3: Sym(1024*(((s0*s1 + 255)//256))) = mul_2 * 4;  mul_2 = None
        floordiv_1: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub // mul_3;  sub = mul_3 = None
        add_2: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_1 + 1;  floordiv_1 = None
        mul_4: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_2 * 4;  add_2 = None
        add_3: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = 0 + mul_4;  mul_4 = None
        add_4: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_3)
        philox_rand_1: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add_4, None, device(type='cuda', index=0), torch.float32);  fwd_seed_1 = add_4 = None
        mul_5: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand_1, mul);  mul = None

        # No stacktrace found for following nodes
        mul_8: Sym(s0*s1) = primals_1 * primals_2
        sub_2: Sym(s0*s1 - 1) = mul_8 - 1
        add_5: Sym(s0*s1 + 256) = mul_8 + 256;  mul_8 = None
        sub_3: Sym(s0*s1 + 255) = add_5 - 1;  add_5 = None
        floordiv_2: Sym(((s0*s1 + 255)//256)) = sub_3 // 256;  sub_3 = None
        mul_9: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv_2;  floordiv_2 = None
        mul_10: Sym(1024*(((s0*s1 + 255)//256))) = mul_9 * 4;  mul_9 = None
        floordiv_3: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub_2 // mul_10;  sub_2 = mul_10 = None
        add_6: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_3 + 1;  floordiv_3 = None
        mul_11: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_6 * 4;  add_6 = None
        add_7: Sym(8*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 8) = add_3 + mul_11;  add_3 = mul_11 = None
        add_8: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_7);  fwd_base_offset_1 = add_7 = None
        return [mul_5, add_8, philox_rand_1, philox_rand, primals_1, primals_2]
 ~~~

The dramatic increase is from the offset calculation, which is now an output node of the fwd and bwd graph. The offset calculations, unfortunately, are complicated and therefore incurs a large number of ops. For inductor, we will have somewhat simpler offset calculation (basically `numel`) 


[ghstack-poisoned]

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this

~~~
 ===== Forward graph 1 =====
 <eval_with_key>.10 class GraphModule(torch.nn.Module):
    def forward(self, fwd_seed_1: i64[], fwd_base_offset_1: i64[], primals_1: Sym(s0), primals_2: Sym(s1), primals_3: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, 0)
        philox_rand: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand, primals_3);  primals_3 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        mul_1: Sym(s0*s1) = primals_1 * primals_2
        sub: Sym(s0*s1 - 1) = mul_1 - 1
        add_1: Sym(s0*s1 + 256) = mul_1 + 256;  mul_1 = None
        sub_1: Sym(s0*s1 + 255) = add_1 - 1;  add_1 = None
        floordiv: Sym(((s0*s1 + 255)//256)) = sub_1 // 256;  sub_1 = None
        mul_2: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv;  floordiv = None
        mul_3: Sym(1024*(((s0*s1 + 255)//256))) = mul_2 * 4;  mul_2 = None
        floordiv_1: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub // mul_3;  sub = mul_3 = None
        add_2: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_1 + 1;  floordiv_1 = None
        mul_4: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_2 * 4;  add_2 = None
        add_3: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = 0 + mul_4;  mul_4 = None
        add_4: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_3)
        philox_rand_1: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add_4, None, device(type='cuda', index=0), torch.float32);  fwd_seed_1 = add_4 = None
        mul_5: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand_1, mul);  mul = None

        # No stacktrace found for following nodes
        mul_8: Sym(s0*s1) = primals_1 * primals_2
        sub_2: Sym(s0*s1 - 1) = mul_8 - 1
        add_5: Sym(s0*s1 + 256) = mul_8 + 256;  mul_8 = None
        sub_3: Sym(s0*s1 + 255) = add_5 - 1;  add_5 = None
        floordiv_2: Sym(((s0*s1 + 255)//256)) = sub_3 // 256;  sub_3 = None
        mul_9: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv_2;  floordiv_2 = None
        mul_10: Sym(1024*(((s0*s1 + 255)//256))) = mul_9 * 4;  mul_9 = None
        floordiv_3: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub_2 // mul_10;  sub_2 = mul_10 = None
        add_6: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_3 + 1;  floordiv_3 = None
        mul_11: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_6 * 4;  add_6 = None
        add_7: Sym(8*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 8) = add_3 + mul_11;  add_3 = mul_11 = None
        add_8: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_7);  fwd_base_offset_1 = add_7 = None
        return [mul_5, add_8, philox_rand_1, philox_rand, primals_1, primals_2]
 ~~~

The dramatic increase is from the offset calculation, which is now an output node of the fwd and bwd graph. The offset calculations, unfortunately, are complicated and therefore incurs a large number of ops. For inductor, we will have somewhat simpler offset calculation (basically `numel`) 


[ghstack-poisoned]

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this

~~~
 ===== Forward graph 1 =====
 <eval_with_key>.10 class GraphModule(torch.nn.Module):
    def forward(self, fwd_seed_1: i64[], fwd_base_offset_1: i64[], primals_1: Sym(s0), primals_2: Sym(s1), primals_3: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, 0)
        philox_rand: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand, primals_3);  primals_3 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        mul_1: Sym(s0*s1) = primals_1 * primals_2
        sub: Sym(s0*s1 - 1) = mul_1 - 1
        add_1: Sym(s0*s1 + 256) = mul_1 + 256;  mul_1 = None
        sub_1: Sym(s0*s1 + 255) = add_1 - 1;  add_1 = None
        floordiv: Sym(((s0*s1 + 255)//256)) = sub_1 // 256;  sub_1 = None
        mul_2: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv;  floordiv = None
        mul_3: Sym(1024*(((s0*s1 + 255)//256))) = mul_2 * 4;  mul_2 = None
        floordiv_1: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub // mul_3;  sub = mul_3 = None
        add_2: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_1 + 1;  floordiv_1 = None
        mul_4: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_2 * 4;  add_2 = None
        add_3: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = 0 + mul_4;  mul_4 = None
        add_4: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_3)
        philox_rand_1: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add_4, None, device(type='cuda', index=0), torch.float32);  fwd_seed_1 = add_4 = None
        mul_5: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand_1, mul);  mul = None

        # No stacktrace found for following nodes
        mul_8: Sym(s0*s1) = primals_1 * primals_2
        sub_2: Sym(s0*s1 - 1) = mul_8 - 1
        add_5: Sym(s0*s1 + 256) = mul_8 + 256;  mul_8 = None
        sub_3: Sym(s0*s1 + 255) = add_5 - 1;  add_5 = None
        floordiv_2: Sym(((s0*s1 + 255)//256)) = sub_3 // 256;  sub_3 = None
        mul_9: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv_2;  floordiv_2 = None
        mul_10: Sym(1024*(((s0*s1 + 255)//256))) = mul_9 * 4;  mul_9 = None
        floordiv_3: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub_2 // mul_10;  sub_2 = mul_10 = None
        add_6: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_3 + 1;  floordiv_3 = None
        mul_11: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_6 * 4;  add_6 = None
        add_7: Sym(8*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 8) = add_3 + mul_11;  add_3 = mul_11 = None
        add_8: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_7);  fwd_base_offset_1 = add_7 = None
        return [mul_5, add_8, philox_rand_1, philox_rand, primals_1, primals_2]
 ~~~

The dramatic increase is from the offset calculation, which is now an output node of the fwd and bwd graph. The offset calculations, unfortunately, are complicated and therefore incurs a large number of ops. For inductor, we will have somewhat simpler offset calculation (basically `numel`) 


[ghstack-poisoned]

ezyang commented:
You're currently doing all of the offset compute with SymInt directly. I don't think this is a good idea, for two reasons: (1) as you can see, when you do SymInt compute, we will maintain Sympy expressions for all of the compute. But you never actually use the computed offset in a size-y context. So the sympy expressions is waste. This can have practical implications, because the more complicated your sympy expressions are, the more likely you will hit some sort of preexisting sympy bug.

We have a standard playbook for this situation: convert the symint to tensor and do the compute on tensor (scalar_to_tensor by @tugsbayasgalan will do the trick)

But there is something even better too: by doing things as tensor compute, you can introduce an operator to compute offsets. And then Inductor can easily override the offset compute by decomposing this operator to something different, if it needs different offsets for its rng.

WDYT.

anijain2305 commented:
@ezyang Yup, I like this idea. Will work on it.

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this


~~~
 ===== Forward graph 1 =====
 <eval_with_key>.10 class GraphModule(torch.nn.Module):
    def forward(self, fwd_seed_1: i64[], fwd_base_offset_1: i64[], primals_1: Sym(s0), primals_2: Sym(s1), primals_3: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, 0)
        philox_rand: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand, primals_3);  primals_3 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        mul_1: Sym(s0*s1) = primals_1 * primals_2
        sub: Sym(s0*s1 - 1) = mul_1 - 1
        add_1: Sym(s0*s1 + 256) = mul_1 + 256;  mul_1 = None
        sub_1: Sym(s0*s1 + 255) = add_1 - 1;  add_1 = None
        floordiv: Sym(((s0*s1 + 255)//256)) = sub_1 // 256;  sub_1 = None
        mul_2: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv;  floordiv = None
        mul_3: Sym(1024*(((s0*s1 + 255)//256))) = mul_2 * 4;  mul_2 = None
        floordiv_1: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub // mul_3;  sub = mul_3 = None
        add_2: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_1 + 1;  floordiv_1 = None
        mul_4: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_2 * 4;  add_2 = None
        add_3: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = 0 + mul_4;  mul_4 = None
        add_4: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_3)
        philox_rand_1: f32[s0, s1] = torch.ops.rngprims.philox_rand.default([primals_1, primals_2], fwd_seed_1, add_4, None, device(type='cuda', index=0), torch.float32);  fwd_seed_1 = add_4 = None
        mul_5: f32[s0, s1] = torch.ops.aten.mul.Tensor(philox_rand_1, mul);  mul = None

        # No stacktrace found for following nodes
        mul_8: Sym(s0*s1) = primals_1 * primals_2
        sub_2: Sym(s0*s1 - 1) = mul_8 - 1
        add_5: Sym(s0*s1 + 256) = mul_8 + 256;  mul_8 = None
        sub_3: Sym(s0*s1 + 255) = add_5 - 1;  add_5 = None
        floordiv_2: Sym(((s0*s1 + 255)//256)) = sub_3 // 256;  sub_3 = None
        mul_9: Sym(256*(((s0*s1 + 255)//256))) = 256 * floordiv_2;  floordiv_2 = None
        mul_10: Sym(1024*(((s0*s1 + 255)//256))) = mul_9 * 4;  mul_9 = None
        floordiv_3: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) = sub_2 // mul_10;  sub_2 = mul_10 = None
        add_6: Sym(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256)))) + 1) = floordiv_3 + 1;  floordiv_3 = None
        mul_11: Sym(4*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 4) = add_6 * 4;  add_6 = None
        add_7: Sym(8*(((s0*s1 - 1)//(1024*(((s0*s1 + 255)//256))))) + 8) = add_3 + mul_11;  add_3 = mul_11 = None
        add_8: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, add_7);  fwd_base_offset_1 = add_7 = None
        return [mul_5, add_8, philox_rand_1, philox_rand, primals_1, primals_2]
 ~~~

The dramatic increase is from the offset calculation, which is now an output node of the fwd and bwd graph. The offset calculations, unfortunately, are complicated and therefore incurs a large number of ops. For inductor, we will have somewhat simpler offset calculation (basically `numel`) 


[ghstack-poisoned]

anijain2305 commented:
@ezyang This is ready for review.

ezyang commented on torch/_prims_common/__init__.py:
```diff
@@ -1801,10 +1800,5 @@ def set_torch_state_tensor(seed, offset):
-        new_offset = offset + relative_offset
-        CUDARngStateHelper.set_torch_state_tensor(seed, new_offset)
+    def set_new_offset(relative_offset):
+        torch.cuda._set_rng_state_offset(relative_offset.item())
```
Does this get traced?

anijain2305 replied:
No, this is post-fix. After the fwd or bwd graph is executed, this is called separately outside to advance the state.

ezyang commented on torch/_prims_common/__init__.py:
```diff
@@ -1784,12 +1784,11 @@ def get_torch_state_as_tuple(fake_mode=None):
 
-        rng_state = torch.cuda.get_rng_state()
+        seed = torch.tensor(torch.cuda.initial_seed())
+        offset = torch.tensor(torch.cuda._get_rng_state_offset())
```
These lines are the ones I mentioned in our 1:1, @anijain2305, instead of allocating a real tensor and then fakeifying it, you should `with fake_mode` and then allocate the tensor within there. In fact, you do not even need the real value, because these are fake tensors! The real value should get read in at some other point.

This is technically a preexisting problem but I think this bit can be simplified a lot if you do it this way.

anijain2305 replied:
Thanks for noticing this. Changed!

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1324,34 +1324,45 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
+        if fw_metadata.is_rng_op_functionalized:
+            # Add the seed and offset to args
+            seed, offset = CUDARngStateHelper.get_torch_state_as_tuple()
+            new_args = [seed, offset, *args]
```
This is bug, you need to clear args here otherwise you will keep the inputs live too long

anijain2305 replied:
IIUC, we call `.clear` inside inductor. But not calling `.clear` here would create an extra reference, rendering `.clear` call inside Inductor ineffective because memory will still be held here. Is that correct?

I changed it other places as well.

Really nice catch!

ezyang replied:
new_args would get cleared, but not args (because `*args` causes a copy)

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this


~~~


[2023-04-24 21:41:37,446] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
 ===== Forward graph 1 =====
 <eval_with_key>.7 class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: i64[], arg1_1: i64[], arg2_1: Sym(s0), arg3_1: Sym(s1), arg4_1: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(arg1_1, 0)
        philox_rand = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        getitem: f32[s0, s1] = philox_rand[0]
        getitem_1: i64[] = philox_rand[1];  philox_rand = None
        add_1: i64[] = torch.ops.aten.add.Tensor(getitem_1, 0);  getitem_1 = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem, arg4_1);  getitem = arg4_1 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        add_2: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_1)
        philox_rand_1 = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add_2, None, device(type='cuda', index=0), torch.float32);  arg2_1 = arg3_1 = arg0_1 = add_2 = None
        getitem_2: f32[s0, s1] = philox_rand_1[0]
        getitem_3: i64[] = philox_rand_1[1];  philox_rand_1 = None
        add_3: i64[] = torch.ops.aten.add.Tensor(add_1, getitem_3);  add_1 = getitem_3 = None
        mul_1: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem_2, mul);  getitem_2 = mul = None

        # No stacktrace found for following nodes
        add_4: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_3);  arg1_1 = add_3 = None
        return (mul_1, add_4)

 ~~~

Each rand op is accompanied by its offset calculation op.


[ghstack-poisoned]

anijain2305 commented:
> This is a dumb nit, but looking at the graph structure now, why doesn't the updated offset get returned from `philox_rand` itself?

Just didn't come to my mind that we could do that. Recent commit cleans this up.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2574,8 +2600,16 @@ def forward(ctx, *deduped_flat_tensor_args):
+            # outputs. However, we need to know the exact boundary to identify
+            # which tensors to be saved for the bwd graph.  num_forward captures
+            # this information.
+            num_forward = num_forward_returns + num_outputs_rng_offset
```
@anijain2305 just curious - do you have an idea of whether export will have a need for functionalizing RNG? Is that something worth preemptively fitting inside of our export schema? (Instead of leaving it out, and having to add it later, I'm not sure what the BC risk there is).

ezyang replied:
I think that some export users will want the functionalized RNG.

anijain2305 replied:
@bdhirsh I am not sure yet. But I think we should assume that they will need functionalized RNG. So, let's review with that in mind.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -551,6 +551,16 @@ def __post_init__(self):
             o.dynamic_dims for o in self.output_info
         )
 
+        self.is_rng_op_functionalized = config.functionalize_rng_ops
```
It's a bit strange that you have to capture the config setting in the collected metadata

ezyang replied:
Is this because the config setting may change later? I guess that makes sense.

anijain2305 replied:
Yes, its because config setting can change later.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1342,37 +1342,50 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
             seed, offset = CUDARngStateHelper.get_torch_state_as_tuple(fake_mode)
-            flat_args = (seed, offset, *flat_args)
+            adjusted_args = (seed, offset, *flat_args)
+            del flat_args  # Remove the reference of inputs from args itself
```
Nope. You gotta *clear* `flat_args`, del'ing it is not enough.

When you call

```
ys = [a, b, c]
f(ys)

def f(xs):
  ...
```

if you del the local xs, the parent frame is still keeping a, b, c alive because ys is a separate reference to the list.

ezyang replied:
Actually, isn't flat_args a bunch of fake tensors? Then this doesn't matter, you don't need to del it at all. (My admonishing above still applies if you were dealing in REAL arguments with the boxed calling convention.)

anijain2305 replied:
This is tuple. So, `clear` wasn't applicable here. I understand the fake tensor part, so this is not important. But, if it was REAL tensors, should I do `list(flat_args).clear()`?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2362,6 +2375,8 @@ def create_functionalized_rng_ops_wrapper(func, args, trace_joint=True):
     # the user.
     fake_mode = detect_fake_mode()
+    if fake_mode is None:
+        fake_mode = FakeTensorMode(shape_env=None)
```
are you sure? Creating a fresh fake tensor mode is typically an antipattern

anijain2305 replied:
No, it doesn't. Good point! I should make it a nullcontext.

ezyang reviewed (approved):
okey dokey

ezyang commented:
@bdhirsh you should make sure you approve of the aot autograd changes

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2545,10 +2569,12 @@ class CompiledFunction(torch.autograd.Function):
                 seed, offset = CUDARngStateHelper.get_torch_state_as_tuple()
-                args = (seed, offset, *args)
+                adjusted_args = (seed, offset, *args)
+                del args  # Remove the reference of inputs from args itself
```
from Ed's comment above - should this be an `args.clear()` instead?

bdhirsh replied:
From Ed's comment above - should this be an `args.clear()` instead?

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this


~~~


[2023-04-24 21:41:37,446] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
 ===== Forward graph 1 =====
 <eval_with_key>.7 class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: i64[], arg1_1: i64[], arg2_1: Sym(s0), arg3_1: Sym(s1), arg4_1: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(arg1_1, 0)
        philox_rand = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        getitem: f32[s0, s1] = philox_rand[0]
        getitem_1: i64[] = philox_rand[1];  philox_rand = None
        add_1: i64[] = torch.ops.aten.add.Tensor(getitem_1, 0);  getitem_1 = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem, arg4_1);  getitem = arg4_1 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        add_2: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_1)
        philox_rand_1 = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add_2, None, device(type='cuda', index=0), torch.float32);  arg2_1 = arg3_1 = arg0_1 = add_2 = None
        getitem_2: f32[s0, s1] = philox_rand_1[0]
        getitem_3: i64[] = philox_rand_1[1];  philox_rand_1 = None
        add_3: i64[] = torch.ops.aten.add.Tensor(add_1, getitem_3);  add_1 = getitem_3 = None
        mul_1: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem_2, mul);  getitem_2 = mul = None

        # No stacktrace found for following nodes
        add_4: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_3);  arg1_1 = add_3 = None
        return (mul_1, add_4)

 ~~~

Each rand op is accompanied by its offset calculation op.


[ghstack-poisoned]

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this


~~~


[2023-04-24 21:41:37,446] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
 ===== Forward graph 1 =====
 <eval_with_key>.7 class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: i64[], arg1_1: i64[], arg2_1: Sym(s0), arg3_1: Sym(s1), arg4_1: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(arg1_1, 0)
        philox_rand = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        getitem: f32[s0, s1] = philox_rand[0]
        getitem_1: i64[] = philox_rand[1];  philox_rand = None
        add_1: i64[] = torch.ops.aten.add.Tensor(getitem_1, 0);  getitem_1 = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem, arg4_1);  getitem = arg4_1 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        add_2: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_1)
        philox_rand_1 = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add_2, None, device(type='cuda', index=0), torch.float32);  arg2_1 = arg3_1 = arg0_1 = add_2 = None
        getitem_2: f32[s0, s1] = philox_rand_1[0]
        getitem_3: i64[] = philox_rand_1[1];  philox_rand_1 = None
        add_3: i64[] = torch.ops.aten.add.Tensor(add_1, getitem_3);  add_1 = getitem_3 = None
        mul_1: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem_2, mul);  getitem_2 = mul = None

        # No stacktrace found for following nodes
        add_4: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_3);  arg1_1 = add_3 = None
        return (mul_1, add_4)

 ~~~

Each rand op is accompanied by its offset calculation op.


[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm!

anijain2305 commented:
@pytorchbot merge

anijain2305 committed:
Update on "[philox_rand] Dynamic shape support"



Extends the functionalization of rng work to Dynamic shapes. An example of the generated graph looks like this


~~~


[2023-04-24 21:41:37,446] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
 ===== Forward graph 1 =====
 <eval_with_key>.7 class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: i64[], arg1_1: i64[], arg2_1: Sym(s0), arg3_1: Sym(s1), arg4_1: f32[s0, s1]):
        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:46, code: a = torch.rand_like(x) * x
        add: i64[] = torch.ops.aten.add.Tensor(arg1_1, 0)
        philox_rand = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add, None, device(type='cuda', index=0), torch.float32);  add = None
        getitem: f32[s0, s1] = philox_rand[0]
        getitem_1: i64[] = philox_rand[1];  philox_rand = None
        add_1: i64[] = torch.ops.aten.add.Tensor(getitem_1, 0);  getitem_1 = None
        mul: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem, arg4_1);  getitem = arg4_1 = None

        # File: /scratch/anijain/work/pytorch/test/test_functionalization_of_rng_ops.py:47, code: a = torch.rand_like(x) * a
        add_2: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_1)
        philox_rand_1 = torch.ops.rngprims.philox_rand.default([arg2_1, arg3_1], arg0_1, add_2, None, device(type='cuda', index=0), torch.float32);  arg2_1 = arg3_1 = arg0_1 = add_2 = None
        getitem_2: f32[s0, s1] = philox_rand_1[0]
        getitem_3: i64[] = philox_rand_1[1];  philox_rand_1 = None
        add_3: i64[] = torch.ops.aten.add.Tensor(add_1, getitem_3);  add_1 = getitem_3 = None
        mul_1: f32[s0, s1] = torch.ops.aten.mul.Tensor(getitem_2, mul);  getitem_2 = mul = None

        # No stacktrace found for following nodes
        add_4: i64[] = torch.ops.aten.add.Tensor(arg1_1, add_3);  arg1_1 = add_3 = None
        return (mul_1, add_4)

 ~~~

Each rand op is accompanied by its offset calculation op.


[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge
